test/GPv2Settlement.test.ts

```diff
@@ -857,6 +867,289 @@ describe("GPv2Settlement", () => {
+              order,
+              // NOTE: Order is treated as fill or kill, and even if the
+              // executed amount is favourable for the user, the settlement
+              // reverts.
+              {
+                receiverBalanceIncreaseAmount: BigNumber.from(
+                  order.buyAmount,
+                ).add(1),
+              },
+            )),
+          ),
+        ).to.be.revertedWith("invalid buy amount");
+      });
+
+      it("reverts when executed buy amount is too low for sell orders", async () => {
+        const order = prepareOrder({
+          kind: OrderKind.SELL,
+          partiallyFillable: true,
+          receiver: receiver.address,
+        });
+
+        await authenticator.connect(owner).addSolver(solver.address);
+        await expect(
+          settlement.connect(solver).executeSingleTradeTest(
+            ...(await prepareTrade(order, {
+              receiverBalanceIncreaseAmount: BigNumber.from(
+                order.buyAmount,
+              ).sub(1),
+            })),
+          ),
+        ).to.be.revertedWith("buy amount too low");
+      });
+
+      it("reverts when executed sell amount is too high for buy orders", async () => {
+        const order = prepareOrder({
+          kind: OrderKind.BUY,
+          partiallyFillable: false,
+          receiver: receiver.address,
+        });
+
+        await authenticator.connect(owner).addSolver(solver.address);
+        await expect(
+          settlement.connect(solver).executeSingleTradeTest(
+            ...(await prepareTrade(order, {
+              transferOutAmount: BigNumber.from(order.sellAmount).add(1),
+            })),
+          ),
+        ).to.be.revertedWith("sell amount too high");
+      });
+    });
+
+    it("rejects expired order", async () => {
+      const order = prepareOrder({
+        kind: OrderKind.SELL,
+        partiallyFillable: false,
+        validTo: 0,
+      });
+
+      await authenticator.connect(owner).addSolver(solver.address);
+      await expect(
+        settlement
+          .connect(solver)
+          .executeSingleTradeTest(...(await prepareTrade(order))),
+      ).to.be.revertedWith("order expired");
+    });
+
+    it("ignores trade executed amount field", async () => {
+      const order = prepareOrder({
+        kind: OrderKind.SELL,
+        partiallyFillable: true,
+      });
+
+      const [tokens, trade, transfers, interactions] = await prepareTrade(
+        order,
+      );
+      trade.executedAmount = ethers.utils.parseEther("1447.42");
+      expect(trade.executedAmount).not.to.equal(order.sellAmount);
+
+      await authenticator.connect(owner).addSolver(solver.address);
+      await settlement
+        .connect(solver)
+        .executeSingleTradeTest(tokens, trade, transfers, interactions);
+
+      expect(
+        await settlement.filledAmount(
+          computeOrderUid(testDomain, order, trader.address),
+        ),
+      ).to.equal(order.sellAmount);
+    });
+  });
+
```

Comment:
I guess technically a lot of test cases would have to be duplicated here (what if allowance manager is not approed, feeDiscount is too high, etc) but it doesn't really seem right. Maybe you can add a test case that verifies/documents the fee amount computation for buy orders.

Reply:
> what if allowance manager is not approed

Then the call to `GPv2AllowanceManager.transferToTarget` would fail, and we have a unit test for this in `GPv2AllowanceManager.test.ts`. The idea would be that this would get caught by a drop in coverage if a refactor happened that would make the transfers take an alternate path.

> feeDiscount is too high

This is a great test! I'll add it.

> Maybe you can add a test case that verifies/documents the fee amount computation for buy orders.

They have a shared path now, since the fees are specified by the input `Trade`. I can add a test case that verifies the computations for a buy order (executed buy/sell/fee amounts since these are different).